Move irrational constants to its own module

fredrikekre opened the pull request.

giordano commented on base/irrationals.jl:
```diff
@@ -142,94 +142,5 @@ end
-log(::Irrational{:e}, x::Number) = log(x)
-
-# align along = for nice Array printing
-function alignment(io::IO, x::Irrational)
```
Was this method removed on purpose?

fredrikekre replied:
No, should have been moved.

Keno commented:
This was discussed in triage and the preferred resolution was the following:
- Delete `eu`
- Make `ℯ` (\mscre) a unicode alias for `e` (i.e. rebase #10612)
- Export `π` and `ℯ` from base by default, put the ascii versions as well as the other math constants in its own module (this PR)
Doesn't all have to be done in this PR, but I wanted to note it down somewhere relevant.

simonbyrne commented:
I think we should probably export `pi` as well.

fredrikekre committed:
move mathematical constants to Base.MathConstants
 - only export pi, π and ℯ from Base
 - remove eu as alias for ℯ
 - add \euler as tab completion for ℯ

fredrikekre commented on base/linalg/lapack.jl:
```diff
@@ -5299,7 +5299,7 @@ for (bdsdc, elty) in
                 q, iq, work, iwork, info)
             chklapackerror(info[])
-            d, e, u, vt, q, iq
+            d, e_, u, vt, q, iq
```
Apparently we returned that mathematical constant `e` from `LAPACK.bdsdc!`. :metal:

andreasnoack replied:
Just checked and `e_` is just garbage after the call to `xbdsdc` so maybe just delete it (and adjust `svd`)

fredrikekre commented:
At time of writing I think this PR now implements the new behaviour. I think the easiest would be to incorporate #10612 in this PR for the deprecation of `eu` etc. What should we name the new module? `Constants`, `MathConsts`, `MathConstants`, `MathematicalConstants` or other suggestions?

giordano commented:
Note that ISO mandates that mathematical constants are to be typeset with roman (upright) font, thus using  `ℯ` for the Euler constant is in contrast with this rule.  Of course, Julia *doesn't have* to follow ISO on this.

simonbyrne commented:
`Irrationals`, since they're all irrational?

fredrikekre commented:
`Irrationals`, since they're all irrational?

In that case we should probably wrap everything regarding irrationals in that module, not just the constants.

andyferris commented:
I actually like that we export ASCII `e` (and definitely `pi`)

KristofferC commented:
Exporting a one letter variable is way too error prone for a general purpose language, in my opinion. I have thrown `e` multiple times, I have had bugs in my code due to missing initializing my own \gamma etc. The cost of someone wanting to use `e` as `exp(1)` having to do an import or using the Unicode variable seems very worthwhile if it avoids putting a one letter variable into all code.

andyferris commented:
Doesn't the exact same argument apply to `I`? (Are there other one letter ASCII exports?)

KristofferC commented:
It sort of does, although I guess it is more rare to use `I` (capital letter) in general code than `e`.  Anyway, I think the path forward here is already decided and most the arguments for and against have been made, so perhaps not necessary to rehash them.

SimonDanisch commented:
I have never used e but use `I` quite a bit and definitely run into problems with it being exported ;) Guess it all depends on the naming style of a particular person!

andyferris commented:
> Anyway, I think the path forward here is already decided and most the arguments for and against have been made, so perhaps not necessary to rehash them.

Is this referring to:

> This was discussed in triage 

?

fredrikekre commented:
`I` would be the last single char base export. We could make that `𝐈` (`\mbfI`) or something, but thats a separate discussion.

giordano commented:
This is a breaking change, because the type of the Euler's number is different.  This can probably be handled in `Compat.jl` by defining the same constant.

ararslan commented:
I must say I still don't like the Unicode alias for `e`. It's too visually similar to ASCII `e` which just makes it confusing. I think we should either keep exporting `e` or put it in the math constants module with no exported alias.

fredrikekre commented:
Could use some guidance on how to deprecate stuff here. Are we OK with just make this breaking? I tried some things e.g.
```julia
export e
e = Irrational{:e}()
Float64(::Irrational{:e}) = (depwarn("e is moved etc...", :Float64); Float64(ℯ))
# ...
```
but then you get a conflict when `using Base.MathConstants`:
```
WARNING: using MathConstants.e in module Main conflicts with an existing identifier.
```
Any way to actually deprecate this?

ararslan commented:
Should be able to use `@deprecate_binding e MathConstants.e`

fredrikekre commented:
Hmmmm, I was sure I tried that, but seems to work. Does not give a very helpful message though:

```julia
julia> e
WARNING: Base.e is deprecated.
  likely near no file:0
ℯ = 2.7182818284590...
```

PallHaraldsson commented:
@Keno "Make ℯ (\mscre)" What does "mscre" stand for anyway? I'm never going to remember this unless I know..

Can we have (also) \euler for "Euler's number"?: https://en.wikipedia.org/wiki/E_%28mathematical_constant%29

https://en.wikipedia.org/wiki/Euler%E2%80%93Mascheroni_constant

fredrikekre commented:
`\mscre`  = `math-script-e`

Keno commented:
mscre stands for "math script e". We can and should add another unicode alias. `\e` has been suggested, but it's annoying, because there's a bunch of other completions that start with `\e..`

fredrikekre commented:
Ok, so lets decide on a name for the module?
This is also a problem:
```julia
julia> e
WARNING: Base.e is deprecated.
  likely near no file:0
ℯ = 2.7182818284590...

julia> e
WARNING: Base.e is deprecated.
  likely near no file:0
ℯ = 2.7182818284590...
```
The warning is rather cryptic, and does not suggest the alternative and also prints more than once. This works though;
```julia
julia> using Base.MathConstants

julia> e
ℯ = 2.7182818284590...
```
Any ideas would be much appreciated.

StefanKarpinski commented on NEWS.md:
```diff
@@ -363,6 +363,15 @@ Deprecated or removed
 
   * `diagm(A::BitMatrix)` has been deprecated, use `diagm(vec(A))` instead ([#23373]).
 
+  * `ℯ` (written as `\mscre<TAB>` or `\euler<TAB>`) is the new default for Euler's
```
What does it mean for this to be "the new default"? Default in what sense?

fredrikekre replied:
The one being exported, and also `typeof(e)` has changed from `Irrational{:e}` to `Irrational{:ℯ}`.

StefanKarpinski replied:
Ok, might be good to clarify that if you're willing!

fredrikekre replied:
Yes, I think my own wording is kinda weird too.
